src/coreclr/tools/r2rdump/CoreDisTools.cs

```diff
@@ -1209,6 +1213,185 @@ private static bool IsAnotherRuntimeFunctionWithinMethod(int rva, RuntimeFunctio
             return false;
         }
 
+        /// <summary>
+        /// Improves disassembler output for RiscV64 by adding comments at the end of instructions.
+        /// </summary>
+        /// <param name="rtf">Runtime function</param>
+        /// <param name="imageOffset">Offset within the image byte array</param>
+        /// <param name="rtfOffset">Offset within the runtime function</param>
+        /// <param name="instruction">Textual representation of the instruction</param>
+        private void ProbeRiscV64Quirks(RuntimeFunction rtf, int imageOffset, int rtfOffset, ref string instruction)
+        {
+            const int InstructionSize = 4;
+            uint instr = BitConverter.ToUInt32(_reader.Image, imageOffset + rtfOffset);
+
+            if (IsRiscV64JalrInstruction(instr))
+            {
+                /*
+                Supported patterns:
+                    auipc
+                    addi
+                    ld
+                    jalr
+            
+                    auipc
+                    ld
+                    jalr
+            
+                    auipc
+                    addi
+                    ld
+                    ld
+                    jalr
+
+                Irrelevant instructions for calle address calculations are skiped.
+                */
+
+                AnalyzeRiscV64Itype(instr, out uint rd, out uint rs1, out int imm);
+                uint register = rs1;
+                int target = imm;
+
+                bool isFound = false;
+                int currentInstrOffset = rtfOffset - InstructionSize;
+                int currentPC = rtf.StartAddress + currentInstrOffset;
+                do
+                {
+                    instr = BitConverter.ToUInt32(_reader.Image, imageOffset + currentInstrOffset);
+
+                    if (IsRiscV64LdInstruction(instr))
+                    {
+                        AnalyzeRiscV64Itype(instr, out rd, out rs1, out imm);
+                        if (rd == register)
+                        {
+                            target = imm;
+                            register = rs1;
+                        }
+                    }
+                    else  if (IsRiscV64AddiInstruction(instr))
+                    {
+                        AnalyzeRiscV64Itype(instr, out rd, out rs1, out imm);
+                        if (rd == register)
+                        {
+                            target =+ imm;
```

Comment:
Nit: Have you seen a pattern which needs addition for `addi` instruction? IMO, it doesn't generate two `addi`s for an address calculation. And it doesn't need both immediate values in `ld` and `addi` because `ld` and `addi` has the same bit size (signed 12 bits) for immediate value.

Reply:
There are patterns which use `addi`. You can find them here, [link](https://github.com/dotnet/runtime/pull/101683#discussion_r1595455992). As you can see, last three examples show patterns which use `addi`. Only the first one uses `ld` to add an immediate value instead of `addi`

Reply:
I mean you don't need `target += imm`. `target = imm` is enough because address is calculated with imm in auipc and one imm in ld or addi in patterns. If I am wrong, please let me know

Reply:
I agree with you that `target = imm` is enough. My explanation for `target += imm` is that i wanted to cover even those patterns which could exist and i didn't meet them. If you think that we shouldn't worry so my for the future I can change it to `target = imm`

Reply:
IMO, we don't need to worry such cases. If there are such cases, it is a bug we need to fix or optimization candidates. :)
If target already has a value when we check `addi`, it means a value is set from other `addi` or `ld` before.
We don't need to generate multiple `addi` for an address calculation.
And `ld` and `addi` have the same offset bits. So `addi` is redundant code in case of that target is set by `ld` before.

However, if you want, you can leave it. That is what I set it a `Nit`.
Thank you.